blog: Legacy projects with unsupported PHP/MySQL versions using DDEV

garvinhicking opened the pull request:
## The Issue

I was encouraged through https://discord.com/channels/664580571770388500/1374723890500337724/1375138791278379084 to contribute my solution as a blog article.

## How This PR Solves The Issue

I have documented my steps to the best of my knowledge.

## Manual Testing Instructions

Run `npm run build` to see the article.

## Automated Testing Overview

Only content, no changes of code.

## Related Issue Link(s)

https://discord.com/channels/664580571770388500/1374723890500337724/1375138791278379084

## Release/Deployment Notes

"Works on my machine". ;)

stasadev commented on public/img/blog/2025/05/museums-victoria-Di7WfLcrJ_I-unsplash.jpg:
Could you please optimized it, 18.3 MB is too much.

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
```diff
@@ -0,0 +1,175 @@
+featureImage:
+  src: /img/blog/2025/05/museums-victoria-Di7WfLcrJ_I-unsplash.jpg
+  alt: Legacy computer museum
+  credit: "Photo by [Museums Victoria](https://unsplash.com/@museumsvictoria?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash) on [Unsplash](https://unsplash.com/photos/gray-mechanical-machine-lot-beside-wall-Di7WfLcrJ_I?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash)"
```
I think stuff like `?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash` is not needed.

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+Note that I configured PHP and MySQL versions that are supported by DDEV for this first:
+
+```
````
Please add language identifiers, such as:

````suggestion
```yaml
````

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+Next I created the very small file `~/legacyphp/.ddev/docker-compose.db.yaml` in the same directory next to `config.yaml`:
+
+```
````
````suggestion
```yaml
````

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+This is done via the file `~/legacyphp/.ddev/docker-compose.php.yaml`:
+
+```
````
````suggestion
```yaml
````

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+A special mount of `~/legacyphp/.ddev/php/` is included so that we can control the `php.ini` configuration, if needed. For example you could disable the OPCache+APC in case you're doing some legacy benchmarking that should not be falsified via caching, I created a very small file `~/legacyphp/.ddev/php/php.ini` file with the contents:
+
+```
````
````suggestion
```ini
````

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+Using `ddev mysql` will connect you to the MySQL 5.5. instance:
+
+```
````
````suggestion
```bash
````

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+To execute PHP with our external PHP docker image, I created the following file in `~/legacyphp/.ddev/apache/apache-site.conf`:
+
+```
````
Not sure if there is something for apache.

garvinhicking replied:
Ah. According to https://shiki.style/languages#bundled-languages `apache` should do the trick, will try.

garvinhicking replied:
looks good to me, it has styling using that markup

stasadev reviewed (commented):
Thank you!

Please check the linter errors https://github.com/ddev/ddev.com/actions/runs/15283249253/job/42987518863?pr=375

And my comments (I didn't made a full review yet.)

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
```diff
@@ -0,0 +1,175 @@
+
+_This guest post is by DDEV community member and [TYPO3](https://typo3.org) contributor [Garvin Hicking](/blog/author/garvin-hicking/)._
+
+In my daily work I am developing [TYPO3](https://typo3.org)-based projects and also contribute to the TYPO3 CMS OpenSource project itself.
```
```suggestion
In my daily work, I develop [TYPO3](https://typo3.org)-based projects and also contribute to the TYPO3 CMS OpenSource project itself.
```

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
```diff
@@ -0,0 +1,175 @@
+I quickly faced three major issues:
+
+- No PHP 5.3 out-of-the-box support from DDEV; it starts with 5.6 as of the time of this writing
+- No MySQL 5.5 support either; it starts with 5.7
```
```suggestion
- No MySQL 5.5 ARM64 support either; it starts with 5.7
```

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,175 @@
+
+## Step 2: Rewire DB
+
+Next I created the very small file `~/legacyphp/.ddev/docker-compose.db.yaml` in the same directory next to `config.yaml`:
+
+```yaml
+services:
+  db:
+    platform: linux/amd64
+    build:
+      args:
+        BASE_IMAGE: ddev/ddev-dbserver-mysql-5.5:v1.24.6
+```
````
This is an interesting workaround you discovered, but it's going to work only until the next `ddev restart`:

```
Failed to start legacyphp: unable to start project legacyphp because the configured database type does not match the current actual database. Please change your database type back to mysql:5.5 and start again, export, delete, and then change configuration and start. To get back to existing type use 'ddev config --database=mysql:5.5' and then you might want to try 'ddev debug migrate-database mysql:8.0', see docs at https://ddev.readthedocs.io/en/stable/users/extend/database-types/
```

I think we can probably insert `platform:linux/amd64` for MySQL 5.5, 5.6 on arm64.

I opened this issue:

- https://github.com/ddev/ddev/issues/7338

But I don't think we're going to implement this fast.

I propose to add a small note on this matter.

garvinhicking replied:
Oooh, shoot. That is unfortunate - I honestly only had to start it up once and left it open for some days, and never issued a restart. I wonder if this makes the whole point of the article moot now, if we do not have a way to re-execute the container.

So I'd possibly scrap the whole article now, if you don't see any other kind of workaround to bypass the database config.yaml assignment/validation. If we had an YAML option like "disable_validation" or so that would massively help of course...

garvinhicking replied:
Hm... I'm a bit afraid this "kills" the lightweightness the solution had before... I'm not so sure this would be a solution we should highlight as a blog entry then?

I would be okay to leave this blog entry dangling, if a ddev release allowing to influence the behaviour could be done at some point. Then we could add an info about the version requirement to the entry.. I am not in a hurry to publish this, but I also don't want to blow up the complexity too much.

stasadev replied:
Let’s put this PR in draft then.

(No new releases this month or next while Randy is bikepacking.)

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
````diff
@@ -0,0 +1,173 @@
+    build:
+      args:
+        BASE_IMAGE: ddev/ddev-dbserver-mysql-5.5:v1.24.6
+```
````
Trick DDEV into thinking it's MySQL 8.0:

````suggestion
    entrypoint: ["sh", "-c", "cp /docker-entrypoint.sh ~/docker-entrypoint.sh; sed -i '157s/.*/if false; then/' ~/docker-entrypoint.sh; sed -i '175s|.*|echo 'mysql_8.0' >/var/lib/mysql/db_mariadb_version.txt|' ~/docker-entrypoint.sh; exec ~/docker-entrypoint.sh"]
```
````

stasadev replied:
This is the file that was edited https://github.com/ddev/ddev/blob/main/containers/ddev-dbserver/files/docker-entrypoint.sh

stasadev commented on src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md:
```diff
@@ -0,0 +1,173 @@
+---
+title: "Using DDEV to spin up a legacy PHP application"
+pubDate: 2025-05-27
+modifiedDate: 2025-05-27
```
```suggestion
pubDate: 2025-05-29
#modifiedDate: 2025-05-29
```

stasadev reviewed (commented):
Please check the `entrypoint` I added.

If we decide to change some logic for MySQL 5.5 and ARM64 in the future, the blog can be modified.

garvinhicking commented:
> Please check the `entrypoint` I added.
> 
> If we decide to change some logic for MySQL 5.5 and ARM64 in the future, the blog can be modified.

That looks like a clever and valid way to circumvent it, and is lightweight. That works for me! I've added a small note to explain this, what do you think? Thanks so much for going the extra mile on it :-)

stasadev reviewed (approved):
> I've added a small note to explain this, what do you think?

I don't see it. Please add it and modify the `pubDate` (let's publish it on Monday or today).

Everything else looks ready.

garvinhicking commented:
> Thank you!
> 
> When you decide to write another blog, use a branch name other than `main`. It lets you work on multiple articles in parallel and makes review easier.

Thanks a lot for the help. Yes, I noticed too I forgot to create a branch first. Will keep it in mind :-)
